### DISCUSSION TITLE
Module_02 - sticking points/questions

### DISCUSSION THREAD

**Discussion by cooperrc** [cooperrc/computational-mechanics#156] [Ideas]
Happy New Year!
I wanted to reach out and see where people are getting stuck or what questions might be popping up as you go through the Module_02 readings + HW. 

I can make a video to try to explain code+concepts, so let me know what you're working on to make it relevant.

**Comment by robert-avena** [comment]
Using the !head command on a local install of jupyter does not seem to work for me:

![image](https://user-images.githubusercontent.com/121516267/210425850-5eccbc72-a984-4326-be13-2c902d755f35.png)

Is there an easy way to fix this such as by installing a package, or should I find another way to look at the column headings such as by opening the csv file?

**Reply by cooperrc** [reply]
Oh man, @c-doll and I we _just_ struggling with this on Windows (Mac+Linux have built-in `head`, but Windows does not). 

Best command we could find was [`more` on DOS](https://www.easydos.com/more.html)

Some background: the `!` is a **bang** operator, it lets you run a command on the computer that's hosting you. In the compmech.uconn.edu server, its a linux container, but if you're running on a local machine it will be Windows/Mac/Linux. Its _super_ helpful to check on _where_ you are and _what_ files exist. 

Some Windows-specific commands to look for missing files (i.e. you get a `File does not exist` error):
_note windows DOS commands are case-insensitive, so `more` ==`MORE` ==`More`_
- `DIR` : list all the files and folders in the current directory
- `MORE`: show the contents of a file
- `CD` : change directory
- `.` : current folder
- `..`: one folder up e.g. if you're working in your Documents\ME3255\HW_02 folder, `.` refers to the `HW_02` folder and `..` refers to the `ME3255` folder

an example, if I'm in the Documents\computational-mechanics\module_02 folder, but I want to list all the files in Documents\computational-mechanics\data, I would run

```bash
! DIR ..\data
```

**Reply by robert-avena** [reply]
Thank you professor, it looks like ```!more``` gives lots of information about the file contents along with the column titles so it should be a usable substitute. 

I did not realize the ```!``` operator runs commands on the host computer! I wonder what else we can do with that.

**Reply by Nadinemm5** [reply]
Im running Jupyter locally as well and got the same error. I was able to use `!head` by opening a jupyter terminal and installing posix with `conda install posix`.

**Comment by pj931** [comment]
I encountered an issue using plt.semilogy and easily coloring the data. It seemed that a lot of matplotlib's colormap functions only work with scatter plots since you are coloring each individual point. I ended up just creating a scatter plot then using 
```python 
plt.yscale('log')
```
I was just wondering if this was what you had intended for us to do and also wanted to drop that line of code/tip here for others that may have been confused or had issues with coloring by designer.

**Reply by cooperrc** [reply]
> creating a scatter plot then using `plt.yscale('log')`

That's a great way to create the scatter plot, the [`plt.scatter`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.scatter.html) allows you to create a list of colors for each x-y point, e.g. 

```python
plt.scatter([0, 1, 2, 3, 4, 5], 
            [0, 1, 4, 9, 16, 25], 
            c = ['k', 'k', 'b', 'r', 'b', 'k'])
```
![image](https://user-images.githubusercontent.com/11371428/210595773-0c29bf2c-a121-421e-95c0-4384efe471c0.png)

Then, you can use the `plt.yscale('log')` to change the y-axis, e.g. 

```python
plt.scatter([0, 1, 2, 3, 4, 5], 
            [0, 1, 4, 9, 16, 25], 
            c = ['k', 'k', 'b', 'r', 'b', 'k'])
plt.yscale('log')
```

Another method is to _loop_ through the data and plot one set at a time, if I want to create the plot I just made:

- colors are black when x < 2 or > 4, labeled 'blackberry'
- colors are blue when x%2== 0, labeled 'blueberry'
- colors are red when x%3 == 0, labeled 'strawberry'

```python
x = np.arange(6)
y = x**2
labels = np.array(['blackberry', 'blackberry', 'blueberry', 'strawberry', 'blueberry', 'blackberry'])
for label in np.unique(labels):
  plt.plot(x[labels == label], y[labels == label],'s', label = label)
plt.legend()
```

![image](https://user-images.githubusercontent.com/11371428/210598694-003061c1-8969-4aef-84d0-35c02c01f701.png)

By default, Matplotlib will create new colors for each line you add to the plot, but you can create a list/array of colors for each label and use that too, 

e.g. 

```python
x = np.arange(6)
y = x**2
labels = np.array(['blackberry', 'blackberry', 'blueberry', 'strawberry', 'blueberry', 'blackberry'])
colors = ['k', 'b', 'r']
for i,label in enumerate(np.unique(labels)):
  plt.plot(x[labels == label], y[labels == label],'s', c=colors[i], label = label)
plt.legend()
```
